tests/multigrid/FullFACPreconditioner.cpp

```diff
@@ -0,0 +1,350 @@
+{
+    // Deallocate the solver state if the solver is already initialized.
+    if (d_is_initialized)
+    {
+        deallocateSolverState();
+    }
+
+    // Setup operator state.
+    d_hierarchy = solution.getPatchHierarchy();
+    d_coarsest_ln = solution.getCoarsestLevelNumber();
+    d_finest_ln = solution.getFinestLevelNumber();
+    int num_comps = solution.getNumberOfComponents();
+
+#if !defined(NDEBUG)
+    TBOX_ASSERT(d_hierarchy == rhs.getPatchHierarchy());
+    TBOX_ASSERT(d_coarsest_ln == rhs.getCoarsestLevelNumber());
+    TBOX_ASSERT(d_finest_ln == rhs.getFinestLevelNumber());
+    TBOX_ASSERT(num_comps == rhs.getNumberOfComponents());
+#endif
+
+    generateDenseHierarchy(d_hierarchy);
+    d_coarsest_ln = 0;
+    d_finest_ln = d_dense_hierarchy->getFinestLevelNumber();
+
+    // We need to allocate vectors on the dense hierarchy
+    d_u = new SAMRAIVectorReal<NDIM, double>(d_object_name + "::u", d_dense_hierarchy, d_coarsest_ln, d_finest_ln);
+    d_f = new SAMRAIVectorReal<NDIM, double>(d_object_name + "::f", d_dense_hierarchy, d_coarsest_ln, d_finest_ln);
+
+    // We can use the same patch indices in u and f as in solution and rhs.
+    for (int comp_num = 0; comp_num < num_comps; ++comp_num)
+    {
+        // Determine variable type
+        Pointer<HierarchyDataOpsReal<NDIM, double>> vop;
+        Pointer<Variable<NDIM>> var = solution.getComponentVariable(comp_num);
+        Pointer<CellVariable<NDIM, double>> cc_var = var;
+        Pointer<SideVariable<NDIM, double>> sc_var = var;
+        Pointer<NodeVariable<NDIM, double>> nc_var = var;
+        Pointer<EdgeVariable<NDIM, double>> ec_var = var;
+        Pointer<FaceVariable<NDIM, double>> fc_var = var;
+        if (cc_var)
+            vop = new HierarchyCellDataOpsReal<NDIM, double>(d_dense_hierarchy, d_coarsest_ln, d_finest_ln);
+        else if (sc_var)
+            vop = new HierarchySideDataOpsReal<NDIM, double>(d_dense_hierarchy, d_coarsest_ln, d_finest_ln);
+        else if (nc_var)
+            vop = new HierarchyNodeDataOpsReal<NDIM, double>(d_dense_hierarchy, d_coarsest_ln, d_finest_ln);
+        else if (ec_var)
+            vop = new HierarchyEdgeDataOpsReal<NDIM, double>(d_dense_hierarchy, d_coarsest_ln, d_finest_ln);
+        else if (fc_var)
+            vop = new HierarchyFaceDataOpsReal<NDIM, double>(d_dense_hierarchy, d_coarsest_ln, d_finest_ln);
+        else
+            TBOX_ERROR("Unknown data centering\n");
+        d_var_op_map[var.getPointer()] = vop;
+        d_u->addComponent(solution.getComponentVariable(comp_num),
+                          solution.getComponentDescriptorIndex(comp_num),
+                          solution.getControlVolumeIndex(comp_num),
+                          vop);
+        d_f->addComponent(rhs.getComponentVariable(comp_num),
+                          rhs.getComponentDescriptorIndex(comp_num),
+                          rhs.getControlVolumeIndex(comp_num),
+                          vop);
```

Comment:
One issue is whether we need to ensure that the control volume needs to specified correctly. This almost surely does not set correct control volumes.

Reply:
I'm not sure we need to set control volumes. Do we ever need to compute norms of solutions? We're preconditioning only, so we're always doing a set number of iterations.